### PR TITLE
Batched Block Jacobi Preconditioner

### PR DESCRIPTION
**Pull request by aggarwal2000**
This PR adds batched block jacobi preconditioner.

### CODE REVIEW

File: reference/test/solver/batch_bicgstab_kernels.cpp [hunk 1]
```diff
@@ -254,8 +254,9 @@ TEST(BatchBicgstab, CoreCanSolveCsrWithoutScaling)
             .with_default_max_iterations(maxits)
             .with_default_residual_tol(tol)
             .with_tolerance_type(gko::stop::batch::ToleranceType::relative)
-            .with_preconditioner(
-                gko::preconditioner::BatchJacobi<T>::build().on(exec))
+            .with_preconditioner(gko::preconditioner::BatchJacobi<T>::build()
```
Comment (pratikvn) [inline reference/test/solver/batch_bicgstab_kernels.cpp:257]: Same question here.Is this change necessary ? Shouldn't the auto-detect work, or are there any convergence issues ?

**Review by pratikvn — COMMENTED** [review]
First part of the review. Have not yet looked at the kernels. One main question is the common storage for all the blocks, which can in the worst case lead to a lot of extraneous storage.

Is the main issue there about having easy access to the data ? Or are there other reasons you chose to do this ?

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+__global__
+    __launch_bounds__(default_block_size) void extract_common_block_pattern_kernel(
+        const int nrows, const int* const __restrict__ sys_row_ptrs,
+        const int* const __restrict__ sys_col_idxs,
+        const gko::size_type num_blocks,
+        const gko::preconditioner::batched_blocks_storage_scheme storage_scheme,
```
Comment (pratikvn) [inline common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:38]:
```suggestion
        const gko::preconditioner::batched_blocks_storage_scheme& storage_scheme,
```

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            }
+        }
+    }
+}
+
+
+template <typename Group, typename ValueType>
+__device__ __forceinline__ int choose_pivot(Group subwarp_grp,
+                                            const int block_size,
+                                            const ValueType* const block_row,
```
Comment (pratikvn) [inline common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc:83]: Can be restrict ? 
```suggestion
                                            const ValueType* const __restrict__ block_row,
```

File: reference/test/solver/batch_richardson_kernels.cpp [hunk 4]
```diff
@@ -132,7 +132,8 @@ TYPED_TEST(BatchRich, SolvesStencilSystemJacobi)
 
     auto r_1 = gko::test::solve_poisson_uniform(
         this->exec, this->solve_fn, this->opts_1, this->sys_1, 1,
-        gko::preconditioner::BatchJacobi<T>::build().on(this->exec));
```
Comment (pratikvn) [inline reference/test/solver/batch_richardson_kernels.cpp:135]: I think for the BatchRichardson solver, it would be good to have a block Jacobi as an inner solver. I think that will be a common use case, so we should make sure to test that.

File: test/preconditioner/batch_jacobi_kernels.cpp [hunk 5]
```diff
@@ -72,72 +72,221 @@ class BatchJacobi : public CommonTestFixture {
...
+        ref_block_jacobi_prec = BJ::build()
+                         .with_max_block_size(max_blk_sz)
+                         .on(ref)->generate(ref_mtx);
+
+        //so that the block pointers are exactly the same for ref and device
+        const int* block_pointers_generated_by_ref = ref_block_jacobi_prec->get_const_block_pointers();
+        const auto num_blocks_generated_by_ref = ref_block_jacobi_prec->get_num_blocks();
+        gko::array<int> block_pointers_for_device(this->exec, num_blocks_generated_by_ref + 1);
+        exec->copy_from(ref.get(), num_blocks_generated_by_ref + 1, block_pointers_generated_by_ref , 
+        block_pointers_for_device.get_data());
```
Comment (pratikvn) [inline test/preconditioner/batch_jacobi_kernels.cpp:118]: Can be simplified: 
```suggestion
        gko::array<int> block_pointers_for_device(this->exec, ref_block_jacobi_prec->get_const_block_pointers() , ref_block_jacobi_prec->get_const_block_pointers() + num_blocks_generated_by_ref + 1);
```

**Review by pratikvn — APPROVED** [review]
Excellent work! LGTM!

File: test/solver/batch_bicgstab_kernels.cpp [hunk 6]
```diff
@@ -148,7 +148,7 @@ TEST_F(BatchBicgstab, StencilSystemLoggerIsCorrect)
 
     auto r_1 = gko::test::solve_poisson_uniform(
         this->exec, this->solve_fn, this->opts_1, this->sys_1, 1,
-        gko::preconditioner::BatchJacobi<value_type>::build().on(this->exec));
+        gko::preconditioner::BatchJacobi<value_type>::build().with_max_block_size(1u).on(this->exec));
```
[the remark this answers is not in the document]
Comment (aggarwal2000) [reply test/solver/batch_bicgstab_kernels.cpp:151]: Changed this to block jacobi.

File: test/preconditioner/batch_jacobi_kernels.cpp (hunk 5 above)
Comment (aggarwal2000) [reply test/preconditioner/batch_jacobi_kernels.cpp]: Done

File: common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc (hunk 2 above)
Comment (aggarwal2000) [reply common/cuda_hip/preconditioner/batch_jacobi_kernels.hpp.inc]: Passing by reference from host to cuda kernel results in an error.

File: reference/test/solver/batch_richardson_kernels.cpp (hunk 4 above)
Comment (aggarwal2000) [reply reference/test/solver/batch_richardson_kernels.cpp]: This now uses block jacobi.

File: reference/test/solver/batch_bicgstab_kernels.cpp (hunk 1 above)
Comment (aggarwal2000) [reply reference/test/solver/batch_bicgstab_kernels.cpp]: Changed this back to block jacobi.

File: include/ginkgo/core/preconditioner/batch_jacobi.hpp [hunk 7]
```diff
@@ -48,10 +48,119 @@ namespace gko {
...
+     * @param block_id the id of the block from the perspective of individual
+     * batch entry
+     *
+     * @return the offset of the block with id: `block_id` within its batch
+     * entry
+     */
+    GKO_ATTRIBUTES size_type
+    get_block_offset(const size_type block_id) const noexcept
+    {
+        return block_id * max_block_size_ * max_block_size_;
```
[the remark this answers is not in the document]
Comment (aggarwal2000) [reply include/ginkgo/core/preconditioner/batch_jacobi.hpp]: Storing variable sized blocks in the same sized matrices allows easy access to the blocks data without really affecting the performance as the extra storage  (in the global memory) is actually never accessed. However, this has been changed now due to the memory issues caused by the large amount of storage.

**Comment by pratikvn** [comment]
format!

**Comment by pratikvn** [comment]
format!

**Comment by pratikvn** [comment]
format!
